File: src/agent_sdk/providers/ollama.py
from __future__ import annotations

import json
from pathlib import Path
from typing import AsyncIterator, Type
from uuid import uuid4

import aiohttp

from agent_sdk.providers.base import Provider
from agent_sdk.types import (
    AgentError,
    Capability,
    ErrorKind,
    Message,
    ModelInfo,
    Response,
    StructuredResponse,
    T,
    Tier,
    parse_json_from_llm,
)


# ##################################################################
# parameter count to tier
# classify a model into a tier based on its parameter count string.
# models with >20B params get FREE_THINKING, everything else FREE_FAST.
# returns FREE_FAST when parameter size is unknown or unparseable.
def _tier_from_param_size(param_size: str) -> Tier:
    if not param_size:
        return Tier.FREE_FAST
    try:
        size_str = param_size.upper().replace(" ", "")
        if size_str.endswith("B"):
            count = float(size_str[:-1])
        elif size_str.endswith("M"):
            count = float(size_str[:-1]) / 1000.0
        else:
            return Tier.FREE_FAST
        return Tier.FREE_THINKING if count > 20.0 else Tier.FREE_FAST
    except (ValueError, AttributeError):
        return Tier.FREE_FAST


# ##################################################################
# classify aiohttp error
# map aiohttp exceptions to the agent-sdk ErrorKind taxonomy.
# connection errors become NOT_AVAILABLE, timeouts become TIMEOUT.
def _classify_aiohttp_error(exc: Exception) -> ErrorKind:
    if isinstance(exc, aiohttp.ServerConnectionError | aiohttp.ClientConnectorError):
        return ErrorKind.NOT_AVAILABLE
    if isinstance(exc, TimeoutError | aiohttp.ServerTimeoutError):
        return ErrorKind.TIMEOUT
    return ErrorKind.INTERNAL


# ##################################################################
# ollama provider
# HTTP provider for a locally-running Ollama instance. communicates
# with the Ollama REST API at base_url. no authentication required.
# a new aiohttp session is created per call — sessions are not safe
# to share across event loops or concurrent long-running tasks.
class OllamaProvider(Provider):
    name = "ollama"

    # ##################################################################
    # init
    # capture the base URL of the Ollama server. defaults to the
    # standard local Ollama address.
    def __init__(self, base_url: str = "http://localhost:11434") -> None:
        self._base_url = base_url.rstrip("/")

    # ##################################################################
    # available
    # probe the Ollama server by fetching the model list. returns True
    # if the server responds with HTTP 200, False otherwise.
    async def available(self) -> bool:
        try:
            async with aiohttp.ClientSession() as session:
                async with session.get(
                    f"{self._base_url}/api/tags",
                    timeout=aiohttp.ClientTimeout(total=5.0),
                ) as resp:
                    return resp.status == 200
        except Exception:
            return False

    # ##################################################################
    # list models
    # fetch the model list from Ollama and convert each entry into a
    # ModelInfo. all ollama models support TEXT and STRUCTURED capabilities.
    # tier is inferred from parameter count when available.
    async def list_models(self) -> list[ModelInfo]:
        try:
            async with aiohttp.ClientSession() as session:
                async with session.get(
                    f"{self._base_url}/api/tags",
                    timeout=aiohttp.ClientTimeout(total=10.0),
                ) as resp:
                    if resp.status != 200:
                        return []
                    data = await resp.json()
        except Exception:
            return []

        models: list[ModelInfo] = []
        for entry in data.get("models", []):
            model_id = entry.get("name", "")
            if not model_id:
                continue
            details = entry.get("details", {})
            param_size = (details.get("parameter_size") or "").strip()
            tier = _tier_from_param_size(param_size)
            display = model_id.split(":")[0].replace("-", " ").replace("_", " ").title()
            if param_size:
                display = f"{display} ({param_size})"
            models.append(
                ModelInfo(
                    provider="ollama",
                    model_id=model_id,
                    display_name=display,
                    capabilities=frozenset({Capability.TEXT, Capability.STRUCTURED}),
                    tier=tier,
                    supports_streaming=True,
                    supports_structured=True,
                    supports_conversation=True,
                    supports_tools=False,
                )
            )
        return models

    # ##################################################################
    # build messages payload
    # convert agent Message objects to the format Ollama expects.
    # system messages are included as-is; others use role+content form.
    def _build_messages(self, messages: list[Message]) -> list[dict]:
        return [{"role": m.role, "content": m.content} for m in messages]

    # ##################################################################
    # complete
    # POST to /api/chat with stream:false. for structured output, the
    # JSON schema is appended as an instruction to the last system message
    # (or a new system message is added). the response text is parsed
    # and validated against the schema when one is provided.
    async def complete(
        self,
        messages: list[Message],
        model: ModelInfo,
        *,
        schema: Type[T] | None = None,
        tools: list[str] | None = None,
        cwd: str | Path | None = None,
        max_turns: int = 1,
        timeout: float = 120.0,
    ) -> Response | StructuredResponse:
        conversation_id = uuid4()
        turn_id = uuid4()

        msg_list = self._build_messages(messages)

        if schema is not None:
            schema_json = json.dumps(schema.model_json_schema(), indent=2)
            instruction = (
                f"\n\nRespond ONLY with valid JSON that matches this schema:\n{schema_json}\n"
                "Do not include any explanation or markdown. Output raw JSON only."
            )
            # append schema instruction to last system message or add new one
            system_indices = [i for i, m in enumerate(msg_list) if m["role"] == "system"]
            if system_indices:
                last_sys = system_indices[-1]
                msg_list[last_sys] = {
                    "role": "system",
                    "content": msg_list[last_sys]["content"] + instruction,
                }
            else:
                msg_list.insert(0, {"role": "system", "content": instruction.strip()})

        payload = {
            "model": model.model_id,
            "messages": msg_list,
            "stream": False,
        }

        try:
            async with aiohttp.ClientSession() as session:
                async with session.post(
                    f"{self._base_url}/api/chat",
                    json=payload,
                    timeout=aiohttp.ClientTimeout(total=timeout),
                ) as resp:
                    if resp.status == 429:
                        raise AgentError(
                            f"Ollama rate limit: {resp.status}",
                            kind=ErrorKind.RATE_LIMIT,
                        )
                    if resp.status >= 400:
                        body = await resp.text()
                        raise AgentError(
                            f"Ollama error {resp.status}: {body}",
                            kind=ErrorKind.INTERNAL,
                        )
                    data = await resp.json()
        except AgentError:
            raise
        except Exception as exc:
            kind = _classify_aiohttp_error(exc)
            raise AgentError(str(exc), kind=kind) from exc

        text = (data.get("message") or {}).get("content") or ""
        usage_data = data.get("usage", {})

        if schema is not None:
            try:
                parsed_json = parse_json_from_llm(text)
                parsed_instance = schema.model_validate(parsed_json)
            except Exception as exc:
                raise AgentError(
                    f"Failed to parse structured response: {exc}",
                    kind=ErrorKind.INVALID_REQUEST,
                ) from exc
            return StructuredResponse(
                text=text,
                model_used=model,
                conversation_id=conversation_id,
                turn_id=turn_id,
                usage=usage_data,
                parsed=parsed_instance,
            )

        return Response(
            text=text,
            model_used=model,
            conversation_id=conversation_id,
            turn_id=turn_id,
            usage=usage_data,
        )

    # ##################################################################
    # stream
    # POST to /api/chat with stream:true. ollama returns newline-delimited
    # JSON objects. each object with a non-empty message.content is yielded
    # as a string chunk. the final object (done:true) is skipped.
    async def stream(
        self,
        messages: list[Message],
        model: ModelInfo,
        *,
        timeout: float = 120.0,
    ) -> AsyncIterator[str]:
        payload = {
            "model": model.model_id,
            "messages": self._build_messages(messages),
            "stream": True,
        }

        try:
            async with aiohttp.ClientSession() as session:
                async with session.post(
                    f"{self._base_url}/api/chat",
                    json=payload,
                    timeout=aiohttp.ClientTimeout(total=timeout),
                ) as resp:
                    if resp.status == 429:
                        raise AgentError(
                            f"Ollama rate limit: {resp.status}",
                            kind=ErrorKind.RATE_LIMIT,
                        )
                    if resp.status >= 400:
                        body = await resp.text()
                        raise AgentError(
                            f"Ollama error {resp.status}: {body}",
                            kind=ErrorKind.INTERNAL,
                        )
                    async for raw_line in resp.content:
                        line = raw_line.decode("utf-8").strip()
                        if not line:
                            continue
                        try:
                            obj = json.loads(line)
                        except json.JSONDecodeError:
                            continue
                        chunk = (obj.get("message") or {}).get("content") or ""
                        if chunk:
                            yield chunk
                        if obj.get("done"):
                            break
        except AgentError:
            raise
        except Exception as exc:
            kind = _classify_aiohttp_error(exc)
            raise AgentError(str(exc), kind=kind) from exc

File: go/provider/ollama.go
// Package provider contains concrete AI provider implementations.
package provider

import (
	"bytes"
	"context"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"strings"
	"time"

	"github.com/google/uuid"

	sdk "github.com/darrenoakey/daz-agent-sdk/go"
)

// titleCase capitalises the first letter of each space-separated word.
// Replaces the deprecated strings.Title without adding an external dependency.
func titleCase(s string) string {
	words := strings.Fields(s)
	for i, w := range words {
		if len(w) > 0 {
			words[i] = strings.ToUpper(w[:1]) + w[1:]
		}
	}
	return strings.Join(words, " ")
}

// tierFromParamSize classifies a model into a tier based on its parameter
// count string. Models with >20B params get FreeThinking, everything else
// FreeFast. Returns FreeFast when parameter size is unknown or unparseable.
func tierFromParamSize(paramSize string) sdk.Tier {
	if paramSize == "" {
		return sdk.TierFreeFast
	}
	s := strings.ToUpper(strings.ReplaceAll(paramSize, " ", ""))
	var count float64
	if strings.HasSuffix(s, "B") {
		if _, err := fmt.Sscanf(s[:len(s)-1], "%f", &count); err != nil {
			return sdk.TierFreeFast
		}
	} else if strings.HasSuffix(s, "M") {
		if _, err := fmt.Sscanf(s[:len(s)-1], "%f", &count); err != nil {
			return sdk.TierFreeFast
		}
		count /= 1000.0
	} else {
		return sdk.TierFreeFast
	}
	if count > 20.0 {
		return sdk.TierFreeThinking
	}
	return sdk.TierFreeFast
}

// classifyHTTPError maps HTTP/network errors to ErrorKind.
func classifyHTTPError(err error) sdk.ErrorKind {
	msg := strings.ToLower(err.Error())
	if strings.Contains(msg, "connection refused") ||
		strings.Contains(msg, "no such host") ||
		strings.Contains(msg, "dial tcp") {
		return sdk.ErrorNotAvailable
	}
	if strings.Contains(msg, "timeout") ||
		strings.Contains(msg, "deadline exceeded") {
		return sdk.ErrorTimeout
	}
	return sdk.ErrorInternal
}

// OllamaProvider communicates with a locally-running Ollama instance
// via its REST API. No authentication required.
type OllamaProvider struct {
	baseURL string
	client  *http.Client
}

// NewOllamaProvider creates a new Ollama provider. If baseURL is empty,
// it defaults to http://localhost:11434.
func NewOllamaProvider(baseURL string) *OllamaProvider {
	if baseURL == "" {
		baseURL = "http://localhost:11434"
	}
	baseURL = strings.TrimRight(baseURL, "/")
	return &OllamaProvider{
		baseURL: baseURL,
		client:  &http.Client{},
	}
}

// Name returns "ollama".
func (o *OllamaProvider) Name() string {
	return "ollama"
}

// Available probes the Ollama server by fetching the model list.
// Returns true if the server responds with HTTP 200.
func (o *OllamaProvider) Available(ctx context.Context) (bool, error) {
	ctx, cancel := context.WithTimeout(ctx, 5*time.Second)
	defer cancel()

	req, err := http.NewRequestWithContext(ctx, http.MethodGet, o.baseURL+"/api/tags", nil)
	if err != nil {
		return false, nil
	}
	resp, err := o.client.Do(req)
	if err != nil {
		return false, nil
	}
	defer resp.Body.Close()
	io.Copy(io.Discard, resp.Body)
	return resp.StatusCode == 200, nil
}

// tagsResponse is the JSON structure returned by /api/tags.
type tagsResponse struct {
	Models []tagModel `json:"models"`
}

type tagModel struct {
	Name    string         `json:"name"`
	Details tagModelDetail `json:"details"`
}

type tagModelDetail struct {
	ParameterSize string `json:"parameter_size"`
}

// ListModels fetches all models from the Ollama server.
func (o *OllamaProvider) ListModels(ctx context.Context) ([]sdk.ModelInfo, error) {
	ctx, cancel := context.WithTimeout(ctx, 10*time.Second)
	defer cancel()

	req, err := http.NewRequestWithContext(ctx, http.MethodGet, o.baseURL+"/api/tags", nil)
	if err != nil {
		return nil, err
	}
	resp, err := o.client.Do(req)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()

	if resp.StatusCode != 200 {
		return nil, fmt.Errorf("ollama /api/tags returned status %d", resp.StatusCode)
	}

	var data tagsResponse
	if err := json.NewDecoder(resp.Body).Decode(&data); err != nil {
		return nil, fmt.Errorf("failed to decode /api/tags response: %w", err)
	}

	var models []sdk.ModelInfo
	for _, entry := range data.Models {
		modelID := entry.Name
		if modelID == "" {
			continue
		}
		paramSize := strings.TrimSpace(entry.Details.ParameterSize)
		tier := tierFromParamSize(paramSize)

		// Build display name: take part before ':', replace - and _ with spaces, title case
		baseName := modelID
		if idx := strings.Index(baseName, ":"); idx >= 0 {
			baseName = baseName[:idx]
		}
		baseName = strings.ReplaceAll(baseName, "-", " ")
		baseName = strings.ReplaceAll(baseName, "_", " ")
		display := titleCase(baseName)
		if paramSize != "" {
			display = fmt.Sprintf("%s (%s)", display, paramSize)
		}

		models = append(models, sdk.ModelInfo{
			Provider:             "ollama",
			ModelID:              modelID,
			DisplayName:          display,
			Capabilities:         []sdk.Capability{sdk.CapabilityText, sdk.CapabilityStructured},
			Tier:                 tier,
			SupportsStreaming:    true,
			SupportsStructured:  true,
			SupportsConversation: true,
			SupportsTools:       false,
		})
	}
	return models, nil
}

// buildMessages converts Message objects to the format Ollama expects.
func buildMessages(messages []sdk.Message) []map[string]string {
	result := make([]map[string]string, len(messages))
	for i, m := range messages {
		result[i] = map[string]string{
			"role":    m.Role,
			"content": m.Content,
		}
	}
	return result
}

// chatRequest is the JSON payload for /api/chat.
type chatRequest struct {
	Model    string              `json:"model"`
	Messages []map[string]string `json:"messages"`
	Stream   bool                `json:"stream"`
}

// chatResponse is the JSON response from /api/chat (non-streaming).
type chatResponse struct {
	Message struct {
		Content string `json:"content"`
	} `json:"message"`
}

// Complete sends messages to Ollama and returns a full response.
func (o *OllamaProvider) Complete(ctx context.Context, messages []sdk.Message, model sdk.ModelInfo, opts sdk.CompleteOpts) (*sdk.Response, error) {
	timeout := opts.Timeout
	if timeout <= 0 {
		timeout = 300.0
	}
	ctx, cancel := context.WithTimeout(ctx, time.Duration(timeout*float64(time.Second)))
	defer cancel()

	payload := chatRequest{
		Model:    model.ModelID,
		Messages: buildMessages(messages),
		Stream:   false,
	}

	body, err := json.Marshal(payload)
	if err != nil {
		return nil, sdk.NewAgentError(fmt.Sprintf("failed to marshal request: %v", err), sdk.ErrorInternal, nil)
	}

	req, err := http.NewRequestWithContext(ctx, http.MethodPost, o.baseURL+"/api/chat", bytes.NewReader(body))
	if err != nil {
		return nil, sdk.NewAgentError(fmt.Sprintf("failed to create request: %v", err), sdk.ErrorInternal, nil)
	}
	req.Header.Set("Content-Type", "application/json")

	resp, err := o.client.Do(req)
	if err != nil {
		kind := classifyHTTPError(err)
		return nil, sdk.NewAgentError(err.Error(), kind, nil)
	}
	defer resp.Body.Close()

	if resp.StatusCode == 429 {
		return nil, sdk.NewAgentError(fmt.Sprintf("Ollama rate limit: %d", resp.StatusCode), sdk.ErrorRateLimit, nil)
	}
	if resp.StatusCode >= 400 {
		respBody, _ := io.ReadAll(resp.Body)
		return nil, sdk.NewAgentError(fmt.Sprintf("Ollama error %d: %s", resp.StatusCode, string(respBody)), sdk.ErrorInternal, nil)
	}

	var chatResp chatResponse
	if err := json.NewDecoder(resp.Body).Decode(&chatResp); err != nil {
		return nil, sdk.NewAgentError(fmt.Sprintf("failed to decode response: %v", err), sdk.ErrorInternal, nil)
	}

	return &sdk.Response{
		Text:           chatResp.Message.Content,
		ModelUsed:      model,
		ConversationID: uuid.New(),
		TurnID:         uuid.New(),
		Usage:          map[string]any{},
	}, nil
}

// Stream sends messages to Ollama and returns a channel of text chunks.
func (o *OllamaProvider) Stream(ctx context.Context, messages []sdk.Message, model sdk.ModelInfo, opts sdk.StreamOpts) (<-chan sdk.StreamChunk, error) {
	timeout := opts.Timeout
	if timeout <= 0 {
		timeout = 300.0
	}
	ctx, cancel := context.WithTimeout(ctx, time.Duration(timeout*float64(time.Second)))

	payload := chatRequest{
		Model:    model.ModelID,
		Messages: buildMessages(messages),
		Stream:   true,
	}

	body, err := json.Marshal(payload)
	if err != nil {
		cancel()
		return nil, sdk.NewAgentError(fmt.Sprintf("failed to marshal request: %v", err), sdk.ErrorInternal, nil)
	}

	req, err := http.NewRequestWithContext(ctx, http.MethodPost, o.baseURL+"/api/chat", bytes.NewReader(body))
	if err != nil {
		cancel()
		return nil, sdk.NewAgentError(fmt.Sprintf("failed to create request: %v", err), sdk.ErrorInternal, nil)
	}
	req.Header.Set("Content-Type", "application/json")

	resp, err := o.client.Do(req)
	if err != nil {
		cancel()
		kind := classifyHTTPError(err)
		return nil, sdk.NewAgentError(err.Error(), kind, nil)
	}

	if resp.StatusCode == 429 {
		resp.Body.Close()
		cancel()
		return nil, sdk.NewAgentError(fmt.Sprintf("Ollama rate limit: %d", resp.StatusCode), sdk.ErrorRateLimit, nil)
	}
	if resp.StatusCode >= 400 {
		respBody, _ := io.ReadAll(resp.Body)
		resp.Body.Close()
		cancel()
		return nil, sdk.NewAgentError(fmt.Sprintf("Ollama error %d: %s", resp.StatusCode, string(respBody)), sdk.ErrorInternal, nil)
	}

	ch := make(chan sdk.StreamChunk)
	go func() {
		defer close(ch)
		defer resp.Body.Close()
		defer cancel()

		decoder := json.NewDecoder(resp.Body)
		for {
			var obj map[string]any
			if err := decoder.Decode(&obj); err != nil {
				if err != io.EOF {
					ch <- sdk.StreamChunk{Err: err}
				}
				return
			}

			msgObj, _ := obj["message"].(map[string]any)
			if msgObj != nil {
				if content, ok := msgObj["content"].(string); ok && content != "" {
					ch <- sdk.StreamChunk{Text: content}
				}
			}

			if done, _ := obj["done"].(bool); done {
				return
			}
		}
	}()

	return ch, nil
}

// Compile-time check that OllamaProvider satisfies Provider.
var _ sdk.Provider = (*OllamaProvider)(nil)
